ENH: add Wright's generalized Bessel function

lorentzenchr opened the pull request:
#### Reference issue
This function is one crucial step towards #11291. See also mailing list thread https://mail.python.org/pipermail/scipy-dev/2020-January/023918.html.

#### What does this implement/fix?
This PR adds Wright's generalized Bessel function as `scipy.special.wright_bessel`, see https://dlmf.nist.gov/10.46.E1.

#### Additional information.
~~This is a pure Python implementation that aims for correctness and not for speed.~~ I did not find any public implementation as a reference.
As it is an entire function, the series representation is always valid. ~~I tried asymptotic expansions for large argument z, without any luck.~~

Edit:
So far, only non-negative values of rho=a, beta=b and z=x are implemented. There are 5 different approaches depending on the ranges of the arguments.

1. Taylor series expansion in x=0 [1], for x <= 1.
    Involves gamma funtions in each term.
2. Taylor series expansion in x=0 [2], for large a.
3. Taylor series in a=0, for tiny a and not too large x.
4. Asymptotic expansion for large x [3, 4].
    Suitable for large x while still small a and b.
5. Integral representation [5], in principle for all arguments.


References:
- [1] https://dlmf.nist.gov/10.46.E1
- [2] P. K. Dunn, G. K. Smyth (2005), Series evaluation of Tweedie exponential
     dispersion model densities. Statistics and Computing 15 (2005): 267-280.
- [3] E. M. Wright (1935), The asymptotic expansion of the generalized Bessel
     function. Proc. London Math. Soc. (2) 38, pp. 257-270.
     https://doi.org/10.1112/plms/s2-38.1.257
- [4] R. B. Paris (2017), The asymptotics of the generalised Bessel function,
     Mathematica Aeterna, Vol. 7, 2017, no. 4, 381 - 406,
     https://arxiv.org/abs/1711.03006
- [5] Y. F. Luchko (2008), Algorithms for Evaluation of the Wright Function for
     the Real Arguments' Values, Fractional Calculus and Applied Analysis 11(1)
     http://sci-gems.math.bas.bg/jspui/bitstream/10525/1298/1/fcaa-vol11-num1-2008-57p-75p.pdf

person142 commented:
I will note that a pure Python implementation is basically a nonstarter in special. Probably best to discuss this on the mailing list before proceeding further here.

rlucas7 commented:
@lorentzenchr this will need unit tests. Given that this calls `scipy.special.iv()` it might help to take a look at the file here:
https://github.com/scipy/scipy/blob/fdf4cd83e0a80a083da3790453a12862b3460793/scipy/special/tests/test_basic.py
and follow a similar format, in fact it probably makes sense to add your test(s) to that file. 

You'll want to consider typical case inputs as well as `Nan` and `inf` inputs to the function in your tests (and code).

rlucas7 commented:
> I will note that a pure Python implementation is basically a nonstarter in special. Probably best to discuss this on the mailing list before proceeding further here.

I Agree.

lorentzenchr commented:
@rlucas7 I'm happy to add tests once it is settled that this PR is not for nothing.

@person142 @rlucas7 Why is a pure python implementation a nonstarter?

My plan:
- Start discussion on scipy dev mailing list
- Make pure python implementation sound and ready
- Cythonize (maybe someone can assist me here, i.e. needs to call `scipy.special.rgamma` in loop)

person142 commented:
> Why is a pure python implementation a nonstarter?

Main reasons are:

- It's too slow
- Lack of typing forces too much boilerplate, e.g. I'm seeing things like this in the 
  implementation:

  ```
  b = 1. * b  # make it at least a float
  ```

- Conditional logic is very painful to express when operating on arrays, writing scalar kernels is much clearer
- With all the ufunc machinery we have it's not significantly harder to write it in C/Cython/Fortran

lorentzenchr commented:
@person142 Thank you very much for your explanations!
I have one good reason against a scalar kernel: speed. The kth element of the series expansion has `rgamma(a*k+b)`. This is independent of `z`. I expect the input a and b to be scalar, but z may be an array. Therefore, one can reuse the values of `rgamma` for all `z`. As `rgamma` is the heaviest element per term, this gives a great speedup for array input `z`.

person142 commented:
> I expect the input a and b to be scalar, but z may be an array.

So this is where we might want to keep the end goal in mind-if we add this as a public function in special then it's going to have to be a ufunc (for API consistency reasons), so `a` and `b` could also be arrays.

On the other hand, if we say "this is just for getting at the Tweedie distributions, where `a` and `b` will be scalars", then we can think about instead making this a private function in special (we do this pretty frequently already), which frees us from the constraints of the public special API.

I was trying to get at something similar in my response on the mailing list:

- public function in special: lots of extra work
- private function in special: much more flexibility, hopefully leading to less work.

lorentzenchr commented:
:thinking: Right now, I like the private special function.

lorentzenchr commented:
@person142 Meanwhile, I think this could start as a private function with limited range for arguments, e.g. all greater equal 0, but still Cython so that it would be easier to make it public.

Currently, I'm working on a stable form for 0<z<=1. Large z seems to be challenging.

rlucas7 commented:
Are you working towards support for complex values ?

 If so, I’d recommend focus on real values only. Tweedie distribution is real valued and I haven’t seen any references to any complex values applications.

> On Jan 10, 2020, at 1:36 PM, Christian Lorentzen <notifications@github.com> wrote:
> 
> ﻿
> @person142 Meanwhile, I think this could start as a private function with limited range for arguments, e.g. all greater equal 0, but still Cython so that it would be easier to make it public.
> 
> Currently, I'm working on a stable form for 0<z<=1. Large z seems to be challenging.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

lorentzenchr commented:
> Are you working towards support for complex values ? If so, I’d recommend focus on real values only. Tweedie distribution is real valued and I haven’t seen any references to any complex values applications.

Exactly, for the moment all values real, and even positive: (a, b, x) >= 0. Makes it a lot easier.

WarrenWeckesser commented:
For what it's worth, here are some functions that use [`mpmath`](http://mpmath.org/).  `wright_bessel` relies on [`mpmath.nsum`](http://mpmath.org/doc/1.1.0/calculus/sums_limits.html#nsum) to compute an approximation to the infinite series, so it suffers from whatever limitation `mpmath.nsum` has.   The function `wright_bessel_rho1` is an alternative that uses the connection between the function and modified Bessel function when rho is 1.
```
import mpmath


def _wright_bessel_term(k, z, rho, beta): 
   numer = mpmath.power(z, k)
   denom = mpmath.gamma(k*rho + beta)*mpmath.factorial(k)
   return numer / denom


def wright_bessel(z, rho, beta):
    """
    Wright's generalized Bessel function.

    Also known as the Bessel-Maitland function.

    Parameter convention corresponds to

        https://dlmf.nist.gov/10.46#E1

    but here the `z` parameter is given first.
    """
    if z == 0:
        return 1 / mpmath.gamma(beta)
    return mpmath.nsum(lambda k: _wright_bessel_term(k, z, rho, beta),
                       [0, mpmath.inf])


def wright_bessel_rho1(x, beta):
    """
    Wright's generalized Bessel function.

    Also known as the Bessel-Maitland function.

    Parameter convention corresponds to

        https://dlmf.nist.gov/10.46#E1

    but here the `x` parameter is given first, and rho is fixed at 1.
    `x` is assumed to be real.
    """
    beta = mpmath.mpf(beta)
    nu = beta - 1
    if x > 0:
        r = mpmath.sqrt(x)
        w = mpmath.besseli(nu, 2*r) / mpmath.power(r, nu)
    elif x < 0:
        r = mpmath.sqrt(-x)
        w = mpmath.besselj(nu, 2*r) / mpmath.power(r, nu)
    else:
        # x == 0
        w = 1 / mpmath.gamma(beta)
    return w
```

For example,
```
In [123]: import mpmath

In [124]: mpmath.mp.dps = 40

In [125]: z = 0.9

In [126]: wright_bessel(z, 1.0, 0.4)
Out[126]: mpf('1.834786964889907297648075016830869147931931')

In [127]: wright_bessel_rho1(z, 0.4)
Out[127]: mpf('1.834786964889907297648075016830869147931931')

In [128]: z = -19.0

In [129]: wright_bessel(z, 1.0, 0.4)
Out[129]: mpf('-0.5596387413411752048288086838416299778273557')

In [130]: wright_bessel_rho1(z, 0.4)
Out[130]: mpf('-0.5596387413411752048288086838416299778273213')
```

WarrenWeckesser commented on scipy/special/wright_bessel.py:
```diff
@@ -0,0 +1,103 @@
+
+    # first some special cases
+    if np.all(z == 0):
+        return 0
```
When z is 0, the value of the function is 1/Γ(b).

lorentzenchr commented:
@WarrenWeckesser Thank you for the MPMATH examples. When I played around with my mpmath implementation, I noticed that `nsum` tries to find convergence every 10 steps. Therefore, I most often set `steps=[100]`.

person142 commented:
> Large z seems to be challenging.

@lorentzenchr yeah, an asymptotic series will almost surely be needed. Feel free to ping me about implementation details if it's not something you're familiar with. Other than that, let me know when this is ready for more review. I'm going to add the "needs work" label so that it stays off my review queue until then.

lorentzenchr commented:
@person142 Thanks for your guidance.
For my use case, it suffices to implement non-negative arguments only `wright_bessel(a, b, x)`. Going for this I'm making progress. There are the typical 3 main domains for this functions:

- small x where one can use the Taylor series (so far implemented for x<=1)
- intermediate x where an integral representation exists, see [paper](http://sci-gems.math.bas.bg/jspui/bitstream/10525/1298/1/fcaa-vol11-num1-2008-57p-75p.pdf) but be aware of misprints. (I have a working implementation, not pushed yet.)
- large x where an asymptotic series makes sense, see [Wright 1935](https://doi.org/10.1112/plms/s2-38.1.257) and [Paris 2017](https://arxiv.org/abs/1711.03006) (I have a working implementation, not pushed yet.)

I have 4 questions right now:
1. We intend to add this function as an internal function. Which precision should I aim for? Only the taylor series gets around 1e-15.
2. How to best choose the limits of the domains?
3. Is it ok to use `scipy.integrate.quad` for the intermediate x-domain? Possibly in Cython?
4. How to write tests (except for special, known arguments), e.g. against values from a mpmath implementation? This function has 3 arguments, so a data grid gets large very quickly.

Would I better address such questions on the mailing list?

person142 commented:
@lorentzenchr 

> Which precision should I aim for?

Ideally relative error `< 1e-13`, but we sometimes go as low as `1e-7`. After that we tend to intentionally start returning nans. Around the zeros of a function (where the condition number is high) it's ok to lose relative accuracy. It might be sensible here to restrict the parameters to some box where accurate computation is possible and raise an exception on the stats side if those limits are exceeded.

> How to best choose the limits of the domains?

That's always a tough question. Sometimes you can prove things about rates of convergence, and that informs what domain you can use. See e.g.

https://github.com/scipy/scipy/blob/master/scipy/special/_hypergeometric.pxd#L65

for an example of that. Sometimes you compute across a wide range of parameters using all methods, plot where the accuracy is high for each method, and use that to figure out your regions. See e.g.

https://github.com/scipy/scipy/blob/master/scipy/special/_precompute/struve_convergence.py

for an example of that. Another technique is computing with multiple methods at runtime and keeping running error estimates, and then returning the result that has the lowest error estimate. Obviously this one really slows down the computation, so it should be more of a last resort.

> Is it ok to use scipy.integrate.quad for the intermediate x-domain

Using `quad` is generally a bad idea because adaptive quadrature is fairly slow. Ideally you pick e.g. a Gaussian quadrature method and precompute the points/weights that will give you the desired precision. (Maybe using a paneling technique where you split the domain into multiple pieces and use Gaussian quadtrature on each piece.) Picking the right quadrature method can be an entire task unto itself.

> This function has 3 arguments, so a data grid gets large very quickly.

Some tricks here:

- Write tests specifically for the boundaries between methods. Things often go wrong in those regions.
- Precompute the Mpmath values and store the results in a data file (we have a whole setup for this). Since Mpmath is often the limiting factor in the test this lets you make the grid bigger.
- For when the grid really is just too big: switch to random sampling.

lorentzenchr commented:
@person142 This PR grew a little larger than hoped for. If you could skim over the code and give some advice before opening it for a formal review, that would be great.

A precomputed data grid via mpmath is still missing.

person142 commented:
@lorentzenchr something I noticed while skimming: instead of including e.g. Sympy derivations as comments, we usually put the code in the `special/_precompute` directory:

https://github.com/scipy/scipy/tree/master/scipy/special/_precompute

That helps us:

- Pull out some common tools
- (If the derivation is complicated) test it
- Easily rerun the code at a later date

lorentzenchr commented:
@person142 It took some time, but I made good progress. The current state is:
- A private function in module `scipy.special` implemented in Cython
- 5 different implementations for different domains of arguments
- New `scipy.special._precompute.wright_bessel.py` for calculation of certain coefficients
- Some basic tests

I'm struggling with the following:
- Add test with grid of values against `mpmath`. As there are 3 arguments, the grid gets large very fast.
- Decide when to use which implementation. And what are the accuracy guarantees?
- Good parametrisation of the integral representation. One paremeter is free to choose which I called `eps`. It interacts with the other parameters and wrongly chosen, the integrand is highly oscillatory and numerical un-integrable.

If you have any further advice or a suggestions how to proceed, input/feeback is very welcome:smirk:

lorentzenchr commented:
Meanwhile, I made good progress on tweaking the domains for the different implementations and to greatly improve numerical accuracy (like the choice of the free `eps` parameter of the integral representation). From my perspective, this is ready for reviews.

Can someone help out with the current CI failure on Azure:
>  test_local[<Data for wright_bessel: wright_bessel>]
>
>AssertionError:  Max |adiff|: 1.57517e-14 Max |rdiff|: 1.0657e-11 Bad results (1 out of 2218) for the following points (in 
output 0):           0.000100000000000002                          1.e-10               2.000000000000022 =>           0.001478067756960386 !=           0.001478067756976138  (rdiff         1.0656969390201442e-11)

I can't reproduce the number 0.001478067756960386 and on my private machine, all tests pass.

lorentzenchr commented:
@person142 I think the label "needs_work" can be removed. A **lot** of work happened and the next logical step, I think, would be a review.

ilayn commented:
@lorentzenchr Done. 

We don't use the MRG directive much so I basically removed it. In the meantime, if you are comfortable with it, could you please rebase to the current master? It's been quite some time since then.

lorentzenchr commented:
@ilayn @person142 @WarrenWeckesser @rlucas7 I know, resources are sparse. May I kindly ask for literally any comment? This would help to keep being motivated. If you think this won't ever be merged, please be honest.

person142 commented on scipy/special/wright_bessel.py:
```diff
@@ -0,0 +1,28 @@
+from ._ufuncs import _wright_bessel
+
+
+def wright_bessel(a, b, x):
```
Looks like you don't need this Python wrapper here; you can export the ufunc directly (which is preferred-it will mean it is also exported in `cython_special`).

lorentzenchr replied:
This means, the function is accessible as `from scipy.special._ufuncs import _wright_bessel`.

person142 replied:
Yes, or if you make the ufunc `wright_bessel` then it will get put in the `__all__` for `_ufuncs`, which will make it available as `special.wright_bessel`.

lorentzenchr replied:
Ok. Now it is exposed as `scipy.special.wright_bessel` without leading underline.

lorentzenchr replied:
Sorry to bother you again, but the import does not work anymore after deleting `scipy/special/wright_bessel.py` an deleting the "_" prefix.
Now, there is only the `wright_bessel.pxd` file, the entry in `functions.json` and in `add_newdocs.py`. I get the following error:
```python
from scipy.special import wright_bessel                                 
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-645a9cc16623> in <module>
----> 1 from scipy.special import wright_bessel

~/github/scipy/scipy/special/__init__.py in <module>
    631 from .sf_error import SpecialFunctionWarning, SpecialFunctionError
    632 
--> 633 from . import _ufuncs
    634 from ._ufuncs import *
    635 

_ufuncs.pyx in init scipy.special._ufuncs()

ModuleNotFoundError: No module named 'scipy.special.wright_bessel'
```
What am I missing here?

lorentzenchr replied:
I got it. It works now.

person142 commented on scipy/special/tests/test_wright_bessel.py:
```diff
@@ -0,0 +1,90 @@
+     [2, 100, 100000, 5.475089686699177e-153, 1e-11]])
+def test_wright_data_grid_failures(a, b, x, phi, accuracy):
+    """Test cases of test_data that do not reach relative accuracy of 1e-11"""
+    assert wright_bessel(a, b, x) == pytest.approx(phi, rel=accuracy)
```
Don't use `pytest.approx`-stick to `numpy.testing` for consistency.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+from . cimport sf_error
+
+# Euler Gamma constant aka Euler Mascheroni constant
+DEF M_EG = 0.57721566490153286060
```
NumPy already exports Euler Mascheroni: https://docs.scipy.org/doc/numpy-1.15.0/reference/c-api.coremath.html#c.NPY_EULER

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+# DEF exp_zero = -745.13321910194117
+
+
+# 1. Taylor series expansion in x=0, for x <= 1
```
I'd recommend making these comments docstrings instead.

lorentzenchr replied:
As those are only helper functions and their docstring is hardly accessible for users, I thought comments are preferable. I'm not opposed to changing, just trying to understand:
Is there a difference in compiled binary size or loading time between commented lines and docstrings?

person142 replied:
I believe there should be no difference. Mostly the docstrings look more "Pythony", which is nice to have even for the low-level code (maybe scares new contributors less). It's also consistent with the other Cython code in the module, but it's not a big deal.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+        double xk_k, res
+
+    xk_k = x**nstart * Gamma(nstart + 1)  # x^k/k!
+    res = xk_k * rgamma(nstart*a + b)
```
Computing a ratio of gammas in this way tends to lead to avoidable overflow-you usually want to do something like use the Lanczos approximation (it's internally available in `special`) to cancel the large factors analytically.

person142 replied:
(You will also see various places using `exp` + `gammaln` for this purpose, but that is strictly less accurate since you can get cancellation when subtracting the gamma terms.)

lorentzenchr replied:
Good hint with the Lanczos approx!
In this case, there is/was a typo, it should read
```python
xk_k = x**nstart * rgamma(nstart + 1)
```
Then it is clear that is is not the ratio of Gammas but a multiplication. Each term  is just `x^k / (k! * Gamma(k*a+b))`.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+        C[1] = 2*M_EG
+        C[2] = 3*M_EG**2 - M_PI**2/2
+        C[3] = 4*M_EG**3 - 2*M_EG*M_PI**2 + 8*M_Z3
+        C[4] = 5*M_EG**4 - 5*M_EG**2*M_PI**2 + 40*M_EG*M_Z3 + M_PI**4/12
```
Better to write out the decimals here (precomputed with mpmath) to ensure they are correct to the last bit. (`gcc` will do precompute the things involving `M_PI` to arbitrary precision for you, but not for `M_EG`.)

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+        res += A[2]*x*(1+x)*(C[1]   + C[2]*b    + C[3]*B[2] + C[4]*B[3])
+        res += A[3]*x*(x**2+3*x+1) * (C[2]      + C[3]*b    + C[4]*B[2])
+        res += A[4]*x*(x**3+6*x**2+7*x+1) *      (C[3]      + C[4]*b)
+        res += A[5]*x*(x**4+10*x**3+25*x**2+15*x+1) *         C[4]
```
A lot of powers of `x` are getting recomputed here. Is there some recurrence relation for the polynomials in `x`? (Like probably right?)

lorentzenchr replied:
Not that I know of or is evident to me. I put precomputation of powers up to x**4.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+        dg = digamma(b)
+        # dg1 = polygamma(1, b)
+        dg1 = zeta(2, b)
+        # res = 1 - a*x*dg + a**2/2*x*(1+x)*(dg**2 - dg1)
```
Why this commented out line?

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+            dg3 =  6 * zeta(4, b)
+            res += a**4/24. * x*(x**3 + 6*x**2 + 7*x + 1) \
+                   *(dg**4 - 6*dg**2*dg1 + 4*dg*dg2 + 3*dg1**2 - dg3)
+        res *= exp(x) * rgamma(b)
```
This could be an overflow risk-could use the Lanczos approximation to combine `exp(x)` and the large part of `rgamma`.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+    C[1] *= (2*a + 1)*(2 + a) - 12*b*(1 + a - b)
+
+    C[2] = C[0] / (1152 * Ap1[2])
+    C[2] *= 144*B[4] - 96*B[3]*(5*a + 1) + 24*B[2]*(20*A[2] + 5*a - 4) \
```
Here (and in other places) I recommend using parens for line continuation and just _using a ton of lines_. e.g. I would write this like

```python
C[2] *= (
    144*B[4]
    - 96*b[3]*(5*a + 1)
    + 24*B[2]*(20*A[2] + 5*a - 4)
    ...
)
```

it gets long, but is clear-each term has a line, so it's easy to see how many terms there are and exactly what they are. It also allows for really big terms because you can just keep nesting; i.e.

```python
blah = (
    1234
    + (
        5678
        + 9012
        + ...
    )
    + ...
)
```

lorentzenchr replied:
I've tried to find a compromise.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+    A[0] = 1.
+    B[0] = 1.
+    Ap1[0] = 1.
+    for k in range(1, 15):
```
Given that we are building a power basis for `a`, it seems likely that parts of this function could be done with `polevl`, which is going to be both faster and more accurate-it's in `special/cephes/polevl.h`.

lorentzenchr replied:
IIUC, `polevl` uses the Horner form, which has mainly advantageous in accuracy. As in the current implementation, we can reuse the stored powers of a and b (in A and B), I think there is no significant performance gain. What we would need is a 2d or 3d polynomial evaluation function:smirk:
Given that the current version is more readable and it would be a larger refactoring, I'd prefer to keep it this way.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+        # roots of laguerre polynomial of order 50
+        # scipy.special.roots_laguerre(50)[0] or
+        # sympy.integrals.quadrature.import gauss_laguerre(50, 16)[0]
+        double *x_laguerre = \
```
Shouldn't need this continuation character if you put the `[` on this line.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+    elif x == 0:
+        return rgamma(b)
+    elif a == 0:
+        return exp(x) * rgamma(b)
```
This could overflow; want to use something like Lanczos to avoid that.

person142 commented on scipy/special/wright_bessel.pxd:
```diff
@@ -0,0 +1,685 @@
+
+
+@cython.cdivision(True)
+cdef inline double wright_bessel_scalar(double a, double b, double x) nogil:
```
Is there a point where we know we're going to start losing accuracy (large `x`, `a`, `b`)? If so it would be good to start returning NaNs in that region instead of inaccurate results.

lorentzenchr replied:
`b >= rgamma_zero` returns nan, also some others overflow correctly. A hard region a<5, large b (10) and large x (>500), see also the `test_wright_data_grid_failures`.
So far, I've aimed for 1e-11 accuracy. After much work, this seems possible with a few exceptions.

person142 reviewed (commented):
So, I'm definitely not going to have time to get into the algorithmic details, but if we have good test coverage then it should be okay. I've instead gone through the code and pointed out places where there could be simplifications/numerical instability issues. I think that if we address those and the tests look good then we should merge.

person142 commented:
Also you can see that a lot of my comments focus on preventing avoidable overflow when computing the ratios of two large terms, often by using the Lanczos approximation (which is in special in `special/cephes/lanczos.c`). Boost has good discussion on this:

https://www.boost.org/doc/libs/1_38_0/libs/math/doc/sf_and_dist/html/math_toolkit/backgrounders/lanczos.html

lorentzenchr commented:
@person142 Thank you so much for your review comments. I'll try to address all of them as soon as I can.

lorentzenchr commented:
@person142 I addressed all of your comments, extended tests a lot and made improvements for extreme arguments. Open to me is, probably, to return more `np.nan` for very difficult arguments.

To have automatic Cython interface, as you proposed, the function name is now without leading `_`. As a consequence, it now a **public special function**.

The failing test is about lines too long in `special/__init__.py`.
Side question: Where to put it? A separate paragraph or under Bessel functions?

person142 commented:
> Where to put it? A separate paragraph or under Bessel functions?

I'd put it in the Bessel functions section.

lorentzenchr commented:
> > Where to put it? A separate paragraph or under Bessel functions?
> 
> I'd put it in the Bessel functions section.

There it is at the moment.

lorentzenchr commented:
@WarrenWeckesser @person142 May I kindly ping you.

person142 reviewed (approved):
Probably the best thing to do is merge and see what happens in the wild.

I'll leave it open for a few days in case someone else wants to comment; ping me again in a few days if I haven't merged it by then.

lorentzenchr commented:
@person142 Gentle ping.

person142 commented:
The lint failure is real-note that it's running a stricter linting check on just your diff (we added that so that we could converge towards cleaner code while not having big refactoring PRs).

lorentzenchr commented:
All green again.

person142 commented:
Awesome, in it goes. Thanks for sticking with it @lorentzenchr!

lorentzenchr commented:
@person142 Hitting the merge button for a +5000 line diff takes some courage/trust:smirk: I hope this function will help in a few places in the ecosytem around scipy.

rgommers commented:
Awesome! @lorentzenchr I added an entry in https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.7.0#scipyspecial-improvements, if there's more to say please feel free to edit.

WarrenWeckesser commented:
Would anyone be interested in added an enlightening example or two to the `wright_bessel` docstring?  We would like to ensure that the number of functions without examples in their docstrings goes down over time (see https://github.com/scipy/scipy/issues/7168).

lorentzenchr commented:
@WarrenWeckesser I opened #13738 and hope that this is what you had in mind with *enlightening*.:smirk:

lorentzenchr commented:
In case someone is interested, I wrote a blog post about this function with a focus on the integral representation and the magic with the free parameter: https://lorentzen.ch/index.php/2024/06/17/a-tweedie-trilogy-part-iii-from-wrights-generalized-bessel-function-to-tweedies-compound-poisson-distribution/
